parameter client takes node interfaces

Karsten1987 opened the pull request:
I had to refactor the constructor of the AsyncParameterClient in order to create an instance of it from a lifecycle node.

The current implementation keeps a handle on the original node pointer: 
https://github.com/ros2/rclcpp/blob/b1ed15ebc7e120b7e4e95df4caaf7f942fccbf15/rclcpp/src/rclcpp/parameter_service.cpp#L29
Why is that the case?

wjwwood commented on rclcpp/src/rclcpp/parameter_client.cpp:
```diff
@@ -24,29 +24,81 @@
 using rclcpp::parameter_client::AsyncParametersClient;
 using rclcpp::parameter_client::SyncParametersClient;
 
+RCLCPP_PUBLIC
```
This does not need to be in the cpp file.

Karsten1987 commented:
After running CI, all tests are actually failing and I need help on this.
In particular, the internal spin never returns and the tests are timing out:

https://github.com/ros2/rclcpp/blob/interface_constructor_for_parameter_client/rclcpp/src/rclcpp/parameter_client.cpp#L419-L420

any ideas?

wjwwood commented:
I re-reviewed the code, but I don't see any obvious issue.

Karsten1987 commented:
In the end, it was a stupid copy & paste mistake. Fixed the typos and rebased.

CI:
[![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3159)](http://ci.ros2.org/job/ci_linux/3159/)
[![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=2516)](http://ci.ros2.org/job/ci_osx/2516/)
[![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3210)](http://ci.ros2.org/job/ci_windows/3210/)
